hikari/internal/ed25519.py

```diff
@@ -0,0 +1,116 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Helper classes used for handling EdDSA cryptography in the interaction server."""
+
+from __future__ import annotations
+
+__all__: typing.Sequence[str] = ["ED25519Key"]
+
+import typing
+
+from pure25519 import ed25519_oop as _pure_ed25519  # type: ignore[import]
+
+
+class ED25519KeyProto(typing.Protocol):
+    """Protocol of a ED25519 key class used for verifying payloads.
+
+    Parameters
+    ----------
+    public_key : builtins.bytes
+        The byte public key this is being initiated for.
+    """
+
+    def __init__(self, public_key: bytes, /) -> None:
+        raise NotImplementedError
+
+    def verify(self, body: bytes, timestamp: bytes, signature: bytes) -> bool:
+        """Verify a interaction payload received from Discord against this public key.
+
+        Parameters
+        ----------
+        body : bytes
+            The interaction payload.
+        signature : bytes
+            Value of the `"X-Signature-Ed25519"` header.
+        timestamp : bytes
+            Value of the `"X-Signature-Timestamp"` header.
+
+        Returns
+        -------
+        bool
+            Whether the provided arguments match this public key.
+        """
+        raise NotImplementedError
+
+
+class SlowED25519Key:
+    """A pure python implementation of `ED25519KeyProto`.
+
+    This relies on a normal dependency and this will always be available.
+    """
+
+    __slots__: typing.Sequence[str] = ("_public_key",)
+
+    def __init__(self, public_key: bytes, /) -> None:
+        self._public_key = _pure_ed25519.VerifyingKey(public_key)
+
+    def verify(self, body: bytes, timestamp: bytes, signature: bytes) -> bool:
+        try:
+            self._public_key.verify(signature, timestamp + body)
+            return True
+
+        except _pure_ed25519.BadSignatureError:
+            return False
+
+
+FastED25519Key: typing.Optional[typing.Type[ED25519KeyProto]]
+"""A C implementation of `ED25519KeyProto`.
+
+Unlike `SlowED25519Key` this depends on a speedup dependency and will be
+`builtins.None` if it's unavailable.
+"""
+
+try:
+    import ed25519 as _ed25519  # type: ignore[import]
+
+    class _FastED25519Key:
+        __slots__: typing.Sequence[str] = ("_public_key",)
+
+        def __init__(self, public_key: bytes, /) -> None:
+            self._public_key = _ed25519.VerifyingKey(public_key)
+
+        def verify(self, body: bytes, timestamp: bytes, signature: bytes) -> bool:
+            try:
+                self._public_key.verify(signature, timestamp + body)
+                return True
+
+            except _ed25519.BadSignatureError:
+                return False
+
+    FastED25519Key = _FastED25519Key
+
+except ImportError:
+    FastED25519Key = None
```

Comment:
This is not the way to do this. There should be a `ED25519Key` class. In the top of the file, you try to import `ed25519`, if it fails, fallback to `pure25519` in the `ED25519Key` class

Reply:
And if we're using that implicit approach how would we test that we're using both implementations properly

Reply:
For what I can see, they seem to do the same thing:
```py
try:
    import ed25519 as _ed25519
    LIB = _ed25519
except ImportError:
    from pure25519 import ed25519_oop as _pure_ed25519
    LIB = _pure_ed25519
    
class ED25519Key:
    """ED25519 key class used for verifying payloads.

    Parameters
    ----------
    public_key : builtins.bytes
        The byte public key this is being initiated for.
    """

    __slots__: typing.Sequence[str] = ("_public_key",)

    def __init__(self, public_key: bytes, /) -> None:
        self._public_key = LIB.VerifyingKey(public_key)

    def verify(self, body: bytes, timestamp: bytes, signature: bytes) -> bool:
        """Verify a interaction payload received from Discord against this public key.

        Parameters
        ----------
        body : bytes
            The interaction payload.
        signature : bytes
            Value of the `"X-Signature-Ed25519"` header.
        timestamp : bytes
            Value of the `"X-Signature-Timestamp"` header.

        Returns
        -------
        bool
            Whether the provided arguments match this public key.
        """
            try:
                self._public_key.verify(signature, timestamp + body)
                return True

            except lib.BadSignatureError:
                return False

Reply:
Still how would we cover both cases with tests, the case of lib a and the case of lib b with a conditional if available check before running

Reply:
You can assume `pure25519` will always be there. Then, in the tests, you can unimport them, mock the imports and see what `LIB` (temp var, dont name it that, pls) is set to

Reply:
> You can assume `pure25519` will always be there. Then, in the tests, you can unimport them, mock the imports and see what `LIB` (temp var, dont name it that, pls) is set to

That's not an integration test that makes sure we're using these different libs properly though, that just tests the behaviour around conditional imports

Reply:
Both libs use a use `VerifyingKey`, `VerifyingKey.verify` and `BadSignatureError`. There is no difference between them. You can also write down some integration tests if you want that mock the imports for both cases. Then you can write tests per lib

Reply:
Is there even a way to mock imports and force them to fail though

Reply:
```py
import mock
import builtins

with mock.patch.object(builtins, "__import__", new=mock.Mock(side_effect=RuntimeError)):
    import test
```

```
Traceback (most recent call last):
  File "/home/davfsa/coding/test-bots/test5.py", line 5, in <module>
    import test
  File "/home/davfsa/coding/hikari/.venv/lib/python3.8/site-packages/mock/mock.py", line 1100, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/home/davfsa/coding/hikari/.venv/lib/python3.8/site-packages/mock/mock.py", line 1104, in _mock_call
    return _mock_self._execute_mock_call(*args, **kwargs)
  File "/home/davfsa/coding/hikari/.venv/lib/python3.8/site-packages/mock/mock.py", line 1161, in _execute_mock_call
    raise effect
RuntimeError
```

Reply:
except you can't target specific imports that way so you're going to have to rely on order and import every other import the module uses which might fuck with future imports before you eh so that you can provide every other import in the mock patch's side effects which sounds excessive, jank and fragile.

Reply:
```py
import mock
import builtins
import contextlib


@contextlib.contextmanager
def mock_import(name, error):
    original = builtins.__import__

    def side_effect(self, *args, **kwargs):
        module = mock.call_args[0][0]  # get the call
        if module == name:
            raise error
        else:
            side_effect.self = self
            return original(self, *args, **kwargs)

    patcher = mock.patch.object(builtins, "__import__", side_effect=side_effect)
    yield patcher.start()
    patcher.stop()


with mock_import("test", RuntimeError) as mock:
    import hikari

    print("hikari imported")
    print(hikari.BotApp)

    import test

    print("this will never get printed")
```
```
hikari imported
<class 'hikari.impl.bot.BotApp'>
Traceback (most recent call last):
  File "/home/davfsa/coding/test-bots/test5.py", line 29, in <module>
    import test
  File "/home/davfsa/coding/hikari/.venv/lib/python3.8/site-packages/mock/mock.py", line 1100, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/home/davfsa/coding/hikari/.venv/lib/python3.8/site-packages/mock/mock.py", line 1104, in _mock_call
    return _mock_self._execute_mock_call(*args, **kwargs)
  File "/home/davfsa/coding/hikari/.venv/lib/python3.8/site-packages/mock/mock.py", line 1167, in _execute_mock_call
    result = effect(*args, **kwargs)
  File "/home/davfsa/coding/test-bots/test5.py", line 13, in side_effect
    raise error
RuntimeError
```

Reply:
I still don't exactly see what the issue is with the original approach though, it's just the class equivalent of what we do in time where we define a standard implementation then try to overwrite that except without redefining a class which would be a mypy and flake8 violation and with an approach that's easier to test but if you don't want the individual classes to be accessible by themselves at runtime as internal detail (tbf this whole thing is internal) then we could just delete them or keep them private or just put a flake8 and mypy ignore on the redefine errors ig

Reply:
Its nicer and removes a lot of duplicate code

Reply:
How is it nicer if it forces you to do a ton of hacky code to test it properly?